Fix excessive synchronization in `contract` on `Dagger.DArray`

mofeing opened the pull request:
Closes #123

jofrevalles reviewed (commented):
Can we add some tests? :)

Todorbsc commented:
> Can we add some tests? :)

Yeah, we should. Currently, we are trying to fix some things in order to make the code work.

mofeing commented:
Updated the code to the last `Tenet` and `Dagger` changes. Also it now calls `Tenet.contract` instead of `OMEinsum.EinCode`.

@jofrevalles I also added some tests.

jofrevalles commented on test/integration/Dagger_test.jl:
```diff
@@ -0,0 +1,60 @@
+                @test collect(parent(contracted_block_tensor)) ≈ parent(contracted_tensor)
+            end
+
+            #     @testset "block-unblock" begin
```
Why is this commented?

mofeing replied:
Because calling `contract` on a `Dagger.DArray`-backed `Tensor` and a non-`DArray` `Tensor` is not yet supported. But I wanted to keep it.

Actually, I think it is a very edgy case: IMO it is reasonable to ask users to convert first all tensors to `Tensor{DArray}`s before calling `contract`.
